src/library/scala/collection/immutable/HashMap.scala

```diff
@@ -764,6 +796,116 @@ private final class BitmapIndexedMapNode[K, +V](
     else new BitmapIndexedMapNode[K, W](dataMap, nodeMap, newContent, originalHashes, size, cachedJavaKeySetHashCode)
   }
 
+  override def mergeInto[V1 >: V](that: MapNode[K, V1], builder: HashMapBuilder[K, V1], shift: Int)(mergef: ((K, V), (K, V1)) => (K, V1)): Unit = that match {
+    case bm: BitmapIndexedMapNode[K, V] =>
+      if (size == 0) {
+        that.buildTo(builder)
+        return
+      } else if (bm.size == 0) {
+        buildTo(builder)
+        return
+      }
+
+      val allMap = dataMap | bm.dataMap | nodeMap | bm.nodeMap
+
+      // minimumIndex is inclusive -- it is the first index for which there is data or nodes
+      val minimumBitPos: Int = Node.bitposFrom(Integer.numberOfTrailingZeros(allMap))
+      // maximumIndex is inclusive -- it is the last index for which there is data or nodes
+      // it could not be exclusive, because then upper bound in worst case (Node.BranchingFactor) would be out-of-bound
+      // of int bitposition representation
+      val maximumBitPos: Int = Node.bitposFrom(Node.BranchingFactor - Integer.numberOfLeadingZeros(allMap) - 1)
+
+      {
+        var bitpos = minimumBitPos
+        var leftIdx = 0
+        var rightIdx = 0
+        var finished = false
+
+        while (!finished) {
```

Comment:
I think we can remove the `finished` variable if we: 
- Replace the `while` condition with `bitpos < maximumBitPos && bitpos > 0`, where the second clause is to avoid the `bitpos = bitpos << 1` overflowing to negative numbers. 
- Simplify the update to `bitpos = bitpos << 1`, without the `if`

Reply:
This is probably a good idea, yeah. I should remove this (anti)pattern in a few places

Reply:
I've gone in now and replaced all these silly finished booleans with `while (index < maxIndex)` style logic, at least in HashMap